sdk/messaging/azservicebus/receiver.go

```diff
@@ -369,38 +363,33 @@ func (r *Receiver) receiveMessagesImpl(ctx context.Context, maxMessages int, opt
 		return messages, nil
 	}
 
-	return r.drainLink(receiver, messages)
+	return r.drainLink(ctx, receiver, messages)
 }
 
 // drainLink initiates a drainLink on the link. Service Bus will send whatever messages it might have still had and
 // set our link credit to 0.
-func (r *Receiver) drainLink(receiver internal.AMQPReceiver, messages []*ReceivedMessage) ([]*ReceivedMessage, error) {
-	receiveCtx, cancelReceive := context.WithCancel(context.Background())
-
+func (r *Receiver) drainLink(ctx context.Context, receiver internal.AMQPReceiver, messages []*ReceivedMessage) ([]*ReceivedMessage, error) {
 	// start the drain asynchronously. Note that we ignore the user's context at this point
 	// since draining makes sure we don't get messages when nobody is receiving.
-	go func() {
-		if err := receiver.DrainCredit(context.Background()); err != nil {
-			tab.For(receiveCtx).Debug(fmt.Sprintf("Draining of credit failed. link will be closed and will re-open on next receive: %s", err.Error()))
+	if err := receiver.DrainCredit(context.Background()); err != nil {
+		tab.For(ctx).Debug(fmt.Sprintf("Draining of credit failed. link will be closed and will re-open on next receive: %s", err.Error()))
 
-			// if the drain fails we just close the link so it'll re-open at the next receive.
-			if err := r.amqpLinks.Close(context.Background(), false); err != nil {
-				tab.For(receiveCtx).Debug(fmt.Sprintf("Failed to close links on ReceiveMessages cleanup. Not fatal: %s", err.Error()))
-			}
+		// if the drain fails we just close the link so it'll re-open at the next receive.
+		if err := r.amqpLinks.Close(context.Background(), false); err != nil {
+			tab.For(ctx).Debug(fmt.Sprintf("Failed to close links on ReceiveMessages cleanup. Not fatal: %s", err.Error()))
 		}
-		cancelReceive()
-	}()
+	}
 
 	// Receive until the drain completes, at which point it'll cancel
 	// our context.
-	// NOTE: That's a gap here where we need to be able to drain _only_ the internally cached messages
-	// in the receiver. Filed as https://github.com/Azure/go-amqp/issues/71
 	for {
-		am, err := receiver.Receive(receiveCtx)
+		am, err := receiver.Prefetched(context.Background())
```

Comment:
Should this use the `ctx` parameter?

Reply:
No, but I should put a comment on there.

Basically, when the user cancels this function we still have some cleanup we need to do or else we'll have bleed over from call to call (the link is basically a buffer of messages). 

So this is a non-interruptible loop that ensures that any messages sitting in the internal message buffer are out of our hands and into the result.

Reply:
Comment added. Let me know if it helps clarify. I also renamed the ctx parameter.